### PR TITLE
client: Fee rate limit setting

### LINKED ISSUE #858: client: fee rate limit setting
The maxFeeRate config item from the server describes the maximum possible fee rate that could be required at match time for swap txns, however the client software should have a user-configurable limit on what max fee rates are acceptable here.

### PR DESCRIPTION
**Pull request by amassarwi**
This diff adds new fee rate threshold wallet setting which will be checked against server `maxFeeRate` when placing and order.
if `maxFeeRate` is greater than the threshold, client throws an error.

Part of #858

### CODE REVIEW

File: client/core/core.go
Comment (chappjc) [inline client/core/core.go:2592]: For clarity in the calling code, it's best to pass what's needed and not more.  That is, 
`func validateFeeRateLimit(feeRateLimit, maxFeeRate uint64) error {` But the function is so trivial I'd be inclined to just do the logic in the caller.  The only thing that warrants the function is the `feeRateLimit != 0` check and it seems that's not possible since the default is non-zero, right?

File: client/core/wallet.go
Comment (chappjc) [inline client/core/wallet.go:30]: All other fees and fee rates are uint64, so let's be consistent.

File: client/asset/dcr/dcr.go
Comment (chappjc) [inline client/asset/dcr/dcr.go:108]: `"feelimit"` => `"feeratelimit"` for clarity

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1406]: Currently `ConfigOpts` works just fine with empty default values, but you are right, using `ConfigOpts` defaults would make it easier to have asset specific defaults, actually that's was my first approach but then users will see defaults in the form instead of empty optional fields so I changed to empty by default and set value in core.

Anyway, will revert that to have asset specific `ConfigOpts` default without optional fields

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:2592]: true, that check was there before adding the default will get red of it

**Comment by amassarwi** [comment]
@chappjc Updated.

Moved default value from Core to ConfigOpts to define asset specific values.
However currently user need to save the wallet settings at least one time for
the default values to be used in xcWallet. 
Will try figure out how to force defaults without saving the wallet settings.

File: client/core/core.go
Comment (buck54321) [inline client/core/core.go:1401]: Roles are mixed here. These are wallet settings. `Core` doesn't deal with wallet settings. This needs to be handled in the **btc** and **dcr** packages.

Have the `ExchangeWallet` reject these in `FundOrder` based on the `asset.Order.DexConfig.MaxFeeRate`

**Review by buck54321 — CHANGES_REQUESTED** [review]
> However currently user need to save the wallet settings at least one time for the default values to be used in xcWallet.

This would require some kind of `Core`-level upgrade, which isn't implemented yet. You can instead interpret zero as default in the `ExchangeWallet`.

File: client/core/core.go
Comment (buck54321) [reply client/core/core.go]: For `fallbackfee` we used atoms/kB to match dcrwallet.

File: client/core/core.go
Comment (buck54321) [reply client/core/core.go]: I agree. This whole thing is best done at the **core** level, but it's a lot more work. Can switch later. Wallet settings are easy.

File: client/core/core.go
Comment (amassarwi) [reply client/core/core.go:1401]: @chappjc preferred it to be on `xcWallet` instead of each asset's `ExchangeWallet` which was the first approach I considered.  The reason was that we need them on `Core`-level when placing the order.

I see your point regarding mixing roles here tho, as we have all other fields logic on ExchangeWallet level.

Comment (chappjc) [reply client/core/core.go:1401]: Only Core deals with this setting though.  I don't see the point of giving this to the ExchangeWallet.

Comment (amassarwi) [reply client/core/core.go:1401]: Looks like moving them to `ExchangeWallet` would solve the blocking issue i'm currently having, we would be able
to store the defaults on `ExchangeWallet` without the user saving the settings, having them Settings only and parsing them on Core level would make forcing defaults a messy task.

Comment (buck54321) [reply client/core/core.go:1401]: It would be nice to implement this in **core**, but we chose to use wallet settings instead since it's easier, so we should implement this in the wallet.

Comment (buck54321) [reply client/core/core.go:1401]: Just as a general rule, `Core` should never try to interpret the `WalletConfig.Settings`.

Comment (amassarwi) [reply client/core/core.go:1401]: yep, using defaults already exists for other fields( e.g feeFallback: https://github.com/decred/dcrdex/blob/master/client/asset/dcr/dcr.go#L447-L451)

Comment (chappjc) [reply client/core/core.go:1401]: If you want, but ExchangeWallet does not use this setting so it is confusing to have to define it in each backend's `Config` struct and then make accessors so Core can get it.  Why can't Core just pull it out of the map?

Comment (chappjc) [reply client/core/core.go:1401]: If this is really a hang up for you guys, we shouldn't use ConfigOpts at all then.  This is completely unrelated to the ExchangeWallet.

Comment (amassarwi) [reply client/core/core.go:1401]: Hrmm, that's what it currently does, Settings doesn't seem to store the defaults, so in the above code:
`feeLimit := walletCfg.Settings["feeratelimit"] \ == ""`

Comment (chappjc) [reply client/core/core.go:1401]: Sure but you can get the defaults from WalletInfo

Comment (amassarwi) [reply client/core/core.go:1401]: I agree that it would be confusing to store it on `ExchangeWallet` and use it only in Core tho,
Let me try to find out if I could somehow pull also the defaults in Core-level

Comment (amassarwi) [reply client/core/core.go:1401]: @chappjc good point, sounds reasonable to me!

Comment (chappjc) [reply client/core/core.go:1401]: Please follow @buck54321's advice though.  I completely defer to him on decisions of wallet config.  I am just concerned that if we make ExchangeWallet handle the setting in it's own Config, then we have to make getter methods on the ExchangeWallets so that Core can get the value, even though ExchangeWallet doesn't use it internally.

Comment (buck54321) [reply client/core/core.go:1401]: If we want to implement this in `Core`, we should, but I'm not looking to use a hybrid approach that breaks important conventions just for convenience. It's not really a wallet setting, but we've chose to make it a wallet setting for now, so let's use it like a wallet setting. It's also easier, I think. The wallet knows it has a value to use. `Core` would have to check every wallet's map for existence. And in principle, we don't control what the asset developer uses that settings key for. We would also be breaking conventions about who defines and interprets settings keys. Before, the wallet settings were for wallet specific settings that the asset developer needed. Now there are wallet settings whose definition and use are dictated by the **core** implementation.

If we choose to implement this in **core** later, we can switch it up. It's no harder to implement this as wallet settings, and it apparently simplifies the task. It appears that the hybrid solutions is a source of complication rather than a convenience anyway.

Comment (buck54321) [reply client/core/core.go:1401]: > I am just concerned that if we make ExchangeWallet handle the setting in it's own Config, then we have to make getter methods on the ExchangeWallets so that Core can get the value, even though ExchangeWallet doesn't use it internally.

Core doesn't need the value, does it?

Comment (chappjc) [reply client/core/core.go:1401]: It's not a wallet setting, but it is an asset-specific setting.  And because this *has to be implemented in Core*, I guess we need to do this the hard way, without having it in the ConfigOpts.

File: client/asset/dcr/dcr.go
Comment (JoeGruffins) [inline client/asset/dcr/dcr.go:112]: Here also.

**Review by JoeGruffins — APPROVED** [review]
Working for me.

**Comment by amassarwi** [comment]
@JoeGruffins Updated all strings to follow 80 chars limit.

File: client/asset/dcr/dcr.go
Comment (buck54321) [inline client/asset/dcr/dcr.go:133]: Need a space at the end.

File: client/asset/btc/btc.go
Comment (buck54321) [reply client/asset/btc/btc.go:114]: [My previous comment](https://github.com/decred/dcrdex/pull/874#discussion_r545429003) was resolved and maybe a little unclear on my position, but there is now a discrepancy between the `fallbackfee` units and the `feeratelimit` units. For consistency, I do think we should use the same units for both.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/asset/btc/btc.go
Comment (chappjc) [reply client/asset/btc/btc.go:114]: I'm aware of the consistency consideration, but I feel strongly that feeratelimit should be integer units like sat/bytes.  The consideration is how these settings are used.  fallbackfee matches the node and RPC settings there it takes e.g. btc/kb, while the limit corresponds to e.g. sat/byte as per server setting.  I would must rather have these settings changable without mental gymnastics than have them be consistent.

Comment (chappjc) [reply client/asset/btc/btc.go:114]: Hmm, I guess I have to correct myself.  `fallbackFeeRate` is converted to integer too.  Ugh, well I *hate* the decimal units and all the prefixed zeros, but changing fallbackfee units is too disruptive, so that's not changing now.  I especially dislike btc/kb for this limit since it is not applied in that format anywhere, but if we insist on these two settings have consistent units, Ok.  However, I do not thing consistency between these two settings is a more powerful consideration that clarity and easy of use.

Comment (amassarwi) [reply client/asset/btc/btc.go]: Updated the message, @buck54321 are you convinced regarding the integer units ?

Comment (chappjc) [reply client/asset/btc/btc.go]: Something to consider: UI could eventually have a units dropdown similar to the Bitcoin Core Qt interface:

image

So it was made that way so user could specify units anyway they like, the question then becomes "what is the natural/native unit to use on the backend?"

Comment (buck54321) [reply client/asset/btc/btc.go]: > I especially dislike btc/kb for feeratelimit since it is not applied in that format anywhere and the server config (maxfeerate) is in unit/byte

Those seem like implementation details. As something that hits the UI, the unit chosen should be the one that most users expect or are comfortable with. I also personally think that atoms/byte is the best unit to use for fees, but atoms/kB was the UI unit chosen by bitcoind and dcrwallet, and even in that screenshot from Bitcoin Qt, the **per kB** part is not selectable. I think we're losing this battle.

> However, I do not think consistency between these two settings is a more powerful consideration than clarity and easy of use.

Consistency is not mutually exclusive with clarity and ease of use, and it's easy to argue that different units for the same thing on the same form is a formula for bad UX. What's clear and easy is typically what's expected or conventional, no?


More than anything, I just don't care to fret about this any more. If you're adamant about atoms/byte, go for it.

Comment (chappjc) [reply client/asset/btc/btc.go]: Well I was actually trying to talk myself out of this hangup with the possiblity for GUI providing a units selector say some point.  If that seems reasonable to implement at some point I no longer have an issue with starting this with with consistent units on the GUI, if somewhat difficult to reconcile with the backend and corresponding units.
See any reason that couldn't be achieved at some point in GUI @buck54321 ?

Comment (buck54321) [reply client/asset/btc/btc.go]: I can see that. Maybe as part of a larger solution to solve our hard-coded `1e8` problem, which will need to be fixed for eth. The backends can define conventional and alternative units, each with their own divisor.

Comment (amassarwi) [reply client/asset/btc/btc.go]: Like the sound of that comprehensive solution 😻

**Comment by amassarwi** [comment]
@buck54321 @chappjc I've changed the fee rate limit field type to float and adjusted the code, it's consistent with
fallback fee now.

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:114]: Maybe my understanding of our conclusion was wrong, but I though we were going to go to BTC/kB and commit to improving our units handling to allow for user selection.

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:121]: Not certain here, but would ~~`to get a confirmation`~~ -> `to be mined` be an improvement?

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:493]: Ah. It is expected to be in BTC/kB. Just need to update the `configOpts` description then.

File: client/asset/btc/btc_test.go
Comment (buck54321) [inline client/asset/btc/btc_test.go:400]: You'll need to do this in **client/asset/ltc/ltc.go** too. Might be worth investigating why no tests caught that.

File: client/asset/dcr/dcr.go
Comment (buck54321) [inline client/asset/dcr/dcr.go:116]: DCR/kB

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:496]: Is it worth doing this in the constructor and storing the value to a field? Fine the way you have it too though.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/asset/btc/btc_test.go
Comment (amassarwi) [reply client/asset/btc/btc_test.go:400]: Added

**Comment by amassarwi** [comment]
@buck54321 updated mate

**Review by buck54321 — APPROVED** [review]

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:496]: You might consider rejecting fee rates `1e-8 <= r < 1e-5 `, since the user is clearly trying to set something, and is probably trying to set a low value, but these will truncate to zero, resulting in the default value being used.

Comment (chappjc) [reply client/asset/btc/btc.go:496]: Good point about rounding to zero.  @amassarwi You'll have to rebase to resolve some conflicts before merge.

Comment (amassarwi) [reply client/asset/btc/btc.go:490]: Done

**Comment by amassarwi** [comment]
@buck54321 @chappjc updated

File: client/asset/btc/btc.go
Comment (chappjc) [reply client/asset/btc/btc.go:490]: Sorry to unresolve this since the PR was so close to done, but I think it just got a little hairier than necessary with the resolution to this that was pushed.

There are two things that I'm not content with now: (1) The new `dcrCfg.FeeRateLimit != 0 && dcrCfg.FeeRateLimit < 1/(smallestUnit/1000)` check could instead be _inferred_ after computing `feesLimitPerByte`, and (2) the error return seems avoidable with the reasonable resolution of rounding up to 1 sat/byte for fee rates <1e-5 DCR/kB.

For dcr, consider:

```go
	feesLimitPerByte := uint64(defaultFeeRateLimit)
	if dcrCfg.FeeRateLimit > 0 {
		feesLimitPerByte = toAtoms(dcrCfg.FeeRateLimit / 1000)
		if feesLimitPerByte == 0 {
			feesLimitPerByte = 1 // round up very low specified rates
		}
	}
```

IMO, we just want to avoid using the default when the specified rate was truncated/rounded to zero when converting units.  I don't have a concern about rounding up (to 1 atom/byte) as a resolution when the specified rate is below 1e-5 DCR/kB.  But I would like to avoid duplicate math with the `smallestUnit` const and the added error return.

Personally, I don't think any user will have a problem with rounding up to the nearest atom/byte, but we're bloating lots of code to add this error return, and adding some error prone and redundant computations.

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:57]: This is a little confusing. I see you're trying to use this in two different places, so I assume the reason this is `1e8` instead of `1e-8` like it says in the doc is to make that easier? Regardless, `1e8` is not `small`. It would be more clear to just hard code the `1e8` and `1e-5` in the two places you're using this new constant, I think.

File: client/asset/btc/btc.go
Comment (chappjc) [reply client/asset/btc/btc.go:490]: @buck54321 round up to 1 {atom,sat}/byte instead of lots more error return handling sit ok with you?

Comment (chappjc) [reply client/asset/btc/btc.go:490]: OK, we just spoke offline and decided that actually it's safest to meet `if feesLimitPerByte == 0 {` with an error instead of rounding up to 1.  But let's clean up the math and the variable name as suggested in the other comment.

Comment (buck54321) [reply client/asset/btc/btc.go:490]: So we did chat a little on matrix. The question on the table is two-fold. 1. Should values lower than `1e-5` be rounded up to `1e-5`, and 2) should it be an error for values < `1e-8`?

I think the safest bet is to assume nothing and error for invalid values < `1e-5`, which are arguably invalid. On the other hand, we could error only for values < `1e-8`, and round values `1e-8 <= r < 1e-5` up to the minimum valid value, `1e-5`, but we are making an assumption about the user's intention in that case, so my preference is to simply error for values `v != 0 && v < 1e-5`.

File: client/asset/btc/btc.go
Comment (buck54321) [inline client/asset/btc/btc.go:469]: Instead of adding an `error` return value here, can you just check the config in `BTCCloneWallet`, which already has the `error`?

File: client/asset/btc/btc.go
Comment (chappjc) [reply client/asset/btc/btc.go:490]: We agree about the handling to error for `v != 0 && v < 1e-5`, although it's best to keep the check bulletproof by simply checking as in my suggested code, but instead of defaulting to 1, return an error.  Small final change.

File: client/asset/btc/btc.go
Comment (chappjc) [reply client/asset/btc/btc.go:469]: That's a good idea too. Although doing it in the caller, you'd have to keep the arguably redundant and error prone math of doing the check before the `toSatoshi( ... /1000)` that would definitively tell you if you actually hit the error condition.  I won't nit pick much more though.  The PR is very close to done and we can follow up.  Just a little cleanup as feels right for now is good with me.

File: client/asset/btc/btc.go
Comment (amassarwi) [reply client/asset/btc/btc.go:490]: 👍 

will do

File: client/asset/btc/btc.go
Comment (amassarwi) [reply client/asset/btc/btc.go]: done, actually no need to use 1e-5 hard coded anywhere with last commit changes.

**Review by chappjc — APPROVED** [review]

**Review by buck54321 — APPROVED** [review]

**Pull request merged**
